### PR TITLE
RTC

### PR DESCRIPTION
**Pull request by cbolles**
Bring in RTC support as a device of the STM32f302r8. Includes getting a struct back as a representation of time as well as the seconds since epoch.

### CODE REVIEW

File: include/EVT/dev/platform/f3xx/f302x8/RTC302x8.hpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    /**
+     * Empty constructor. Will startup the RTC.
+     */
+    RTCf302x8();
+
+    /**
+     * Get the current time as determined by the real time clock.
+     *
+     * @param[out] time The time struct to populate
```
Comment (ddp5730) [inline include/EVT/dev/platform/f3xx/f302x8/RTC302x8.hpp:28]: I haven't seen this `[out]` or `[in]` syntax before.  Is this used by sphinx?

File: src/dev/platform/f3xx/f302x8/RTCf302x8.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    uint32_t y = ts.year;
+    uint32_t m = ts.month;
+    uint32_t d = ts.day;
+    uint32_t h = ts.hour;
+    uint32_t min = ts.minute;
+    uint32_t s = ts.second;
+
+
+    // Get number of days sinc epoch
```
Comment (ddp5730) [inline src/dev/platform/f3xx/f302x8/RTCf302x8.cpp:70]: Typo

File: src/dev/platform/f3xx/f302x8/RTCf302x8.cpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    y -= m <= 2;
+    const uint32_t era = (y >= 0 ? y : y-399) / 400;
+    const uint32_t yoe = static_cast<uint32_t>(y - era * 400);
+    const uint32_t doy = (153*(m + (m > 2 ? -3 : 9)) + 2)/5 + d-1;
+    const uint32_t doe = yoe * 365 + yoe/4 - yoe/100 + doy;
+    uint32_t time = era * 146097 + static_cast<int32_t>(doe) - 719468;
```
Comment (ddp5730) [inline src/dev/platform/f3xx/f302x8/RTCf302x8.cpp:76]: Ngl, I have no clue at all what is happening here.

**Review by ddp5730 — CHANGES_REQUESTED** [review]
It looks great.  I'd just like to clear up why the implementation of RTCf302x8::getTime is so convoluted.  The implementation just isn't very readable or intuitive.

File: include/EVT/dev/platform/f3xx/f302x8/RTC302x8.hpp (hunk 1 above)
Comment (cbolles) [reply include/EVT/dev/platform/f3xx/f302x8/RTC302x8.hpp:28]: Its actually a C/C++ doc convention. In means the parameter is used as output. Out means the parameter is being used as an output. So in this example the time struct is actually storing the output of this function.

File: src/dev/platform/f3xx/f302x8/RTCf302x8.cpp (hunk 2 above)
Comment (cbolles) [reply src/dev/platform/f3xx/f302x8/RTCf302x8.cpp]: Good catch, should be fixed now

File: src/dev/platform/f3xx/f302x8/RTCf302x8.cpp (hunk 3 above)
Comment (cbolles) [reply src/dev/platform/f3xx/f302x8/RTCf302x8.cpp:78]: That's fair. Its a strange algorithm that I don't fully understand either. I've linked to the original source. I did verify that the algorithm does indeed work by comparing the output to a known good converter.

Comment (cbolles) [reply src/dev/platform/f3xx/f302x8/RTCf302x8.cpp:78]: That's fair. Its a strange algorithm that I don't fully understand either. I've linked to the original source. I did verify that the algorithm does indeed work by comparing the output to a known good converter.

File: include/EVT/dev/platform/f3xx/f302x8/RTC302x8.hpp (hunk 1 above)
Comment (zsteig) [reply include/EVT/dev/platform/f3xx/f302x8/RTC302x8.hpp:28]: > Its actually a C/C++ doc convention.

Will we be adopting this convention to the rest of the codebase?

File: include/EVT/utils/time.hpp [hunk 4]
```diff
@@ -20,5 +20,35 @@ namespace EVT::core::time
...
+    /**
+     * Struct representing a timestamp.
+     */
+    struct TimeStamp {
+        /**
+         * The current year
+         */
+        uint16_t year;
+        /**
+         * The current month (0-12)
```
Comment (zsteig) [inline include/EVT/utils/time.hpp:33]: Shouldn't this be (1-12)? Or (0-11)?

File: include/EVT/dev/platform/f3xx/f302x8/RTC302x8.hpp (hunk 1 above)
Comment (cbolles) [reply include/EVT/dev/platform/f3xx/f302x8/RTC302x8.hpp:28]: Ideally yes, just to clear up if an argument is used as an input, or if the caller should expect the parameter is store the output of the call.

File: include/EVT/utils/time.hpp (hunk 4 above)
Comment (cbolles) [reply include/EVT/utils/time.hpp:33]: Lol yes, good catch, 1-12. I'll fix that now

File: src/dev/platform/f3xx/f302x8/RTCf302x8.cpp (hunk 3 above)
Comment (ddp5730) [reply src/dev/platform/f3xx/f302x8/RTCf302x8.cpp:78]: Fair enough.  I just wanted to call attention to it.

**Review by ddp5730 — APPROVED** [review]

**Review by zsteig — APPROVED** [review]

**Pull request merged**